### PR TITLE
Making sure that the resolved DNS lists get used in full

### PR DESCRIPTION
**Pull request by gezero**
Signed-off-by: Jiri Peinlich <jiri.peinlich@gmail.com>

## PR description

Previously we would always process the list of DNS resolved peers from the beginning. We would
at most only try `maxConnections` valid elements from the list and never try more. So the each
time we might be trying the same 25 peers, we connect with them and then drop them. This PR
firstly shuffles the peers list each time before applying it and then also takes from the list for
as long as there are free peers slots available as opposed to guess the limit upfront and hope
that all the peers connect.

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java [hunk 1]
```diff
@@ -376,6 +377,7 @@ void attemptPeerConnections() {
   public Stream<DiscoveryPeer> streamDiscoveredPeers() {
     List<DiscoveryPeer> peers = dnsPeers.get();
     if (peers != null) {
+      Collections.shuffle(peers);
       return Stream.concat(peerDiscoveryAgent.streamDiscoveredPeers(), peers.stream());
     }
     return peerDiscoveryAgent.streamDiscoveredPeers();
```
Comment (matkt) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java:383]: Maybe we can also shuffle this part ?

**Comment by gezero** [comment]
Unfortunately streams don't have a convenient method to shuffle them. Let me add a unit test to cover the `takewhile`

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java (hunk 1 above)
Comment (gezero) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java:383]: Hmm, unfortunately streams don't have a convenient method to shuffle the elements.

**Comment by gezero** [comment]
I am going to add a test covering the `takeWhile` part

**Comment by gezero** [comment]
I added a test to check the takeWhile

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java (hunk 1 above)
Comment (gezero) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java:383]: Shuffling a stream goes against the "lazy" nature of a stream I would say. To shuffle it you would need to process all elements at least once. So you need to collect, shuffle, and convert result to another stream.

**Review by jflo — APPROVED** [review]

**Review by matkt — APPROVED** [review]
LGTM

**Pull request merged**